File: quickstart/src/main/java/cn/wjhub/java/state/job/StateFlink.java
package cn.wjhub.java.state.job;


import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * @author 张文军
 * @Description:
 * @Company:南京农业大学工学院
 * @version:1.0
 * @date 2021/1/180:48
 */
public class StateFlink {
    public static void main(String[] args) throws Exception {
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(new Configuration());

        final DataStreamSource<String> dataStreamSource = env.socketTextStream("localhost", 9999);


        dataStreamSource.setBufferTimeout(1);

        /*final SingleOutputStreamOperator<String> flatmap = dataStreamSource.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public void flatMap(String value, Collector<String> out) throws Exception {
                final String[] ss = value.replaceAll("\\W", " ").split(" ");
                for (String s : ss) {
                    out.collect(s);
                }
            }
        });

        final SingleOutputStreamOperator<Tuple2<String, Integer>> map = flatmap.map(new MapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public Tuple2<String, Integer> map(String value) throws Exception {
                return new Tuple2<>(value, 1);
            }
        });*/

        final SingleOutputStreamOperator<Tuple2<String, Integer>> map = dataStreamSource.flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public void flatMap(String value, Collector<Tuple2<String, Integer>> out) throws Exception {
                final String[] split = value.replaceAll("\\W", " ").split(" ");
                for (String s : split) {
                    out.collect(new Tuple2<>(s, 1));
                }
            }
        });
        final SingleOutputStreamOperator<Tuple2<String, Integer>> reduce = map.keyBy(0).reduce(new ReduceFunction<Tuple2<String, Integer>>() {
            @Override
            public Tuple2<String, Integer> reduce(Tuple2<String, Integer> value1, Tuple2<String, Integer> value2) throws Exception {
                return new Tuple2<>(value1.f0, value2.f1 + value1.f1);
            }
        });

        reduce.print();
        System.out.println(env.getExecutionPlan());

        env.execute("JavaJob");

    }
}

File: quickstart/src/main/scala/cn/wjhub/scala/StateFlink.scala
package cn.wjhub.scala

import org.apache.flink.api.common.functions.FlatMapFunction
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

import scala.util.Random


/**
 * @Description:
 * @author 张文军
 * @Company:南京农业大学工学院
 * @version:1.0
 * @date 2021/1/109:18
 */
object StateFlink {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(new Configuration())
    val ds = env.socketTextStream("localhost", 9000)

    val data: DataStream[(String, Int, Long)] = ds.

      flatMap(
        _
          .replaceAll("\\W", " ")
          .split(" ")
          .filter(_.nonEmpty)
      )
      .map((_, 1, System.currentTimeMillis() + Random.nextInt()))
      .keyBy(0)
      .reduce((t1, t2) => {
        (t1._1, t1._2 + t2._2, t2._3)
      })


    val wordMap: DataStream[(String, Int)] = ds.flatMap(new FlatMapFunction[String, (String, Int)] {
      override def flatMap(value: String, out: Collector[(String, Int)]) = {
        val s: Array[String] = value.replaceAll("\\W", " ").split(" ")
        for (s1 <- s) {
          out.collect(s1, 1)
        }
      }
    })

    wordMap.keyBy(0).reduce((w1, w2) => (w1._1, w1._2 + w2._2)).print()


    data.print("data")

    env.execute("stateTest")


  }
}



